### PR TITLE
[script.embuary.helper] 1.2.22

### PR DESCRIPTION
**Pull request by sualfred**
### Description

* copy skin nodes function
* get path stats function
* fixed unwanted logging

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: script.embuary.helper/resources/lib/utils.py [hunk 1]
```diff
@@ -127,6 +127,26 @@ def textviewer(params):
     DIALOG.textviewer(headertxt, bodytxt)
 
 
+def copyskinnodes(params):
```
Comment (enen92) [inline script.embuary.helper/resources/lib/utils.py]: Don't think this is allowed

Comment (sualfred) [reply script.embuary.helper/resources/lib/utils.py]: Why shouldn't this be allowed? The library node editor does exactly the same. And the wiki is showing a howto how to create own nodes.
The problem is, that .xml nodes aren't working outside of the library path.

Comment (sualfred) [reply script.embuary.helper/resources/lib/utils.py]: If it's a problem that I move them to
* %USERDATA%/library/%SKINNAME%/node.xml
instead of
* %USERDATA%/library/video/%SKINNAME%/node.xml
then I could change that. I only thought this would be an much more organized and clearer to users.

Background context why skinners are using direct xml nodes in same cases -> .xml nodes overwrite the window history. Otherwise a home button link "all tv show titles" for example could lead into a episode listing instead to the titles if MyVideoNav.xml was left without wallking the whole way back before. The same applies to all other nodes that are stacked.
By using a node you can directly and correctly point to a path without any hassle. 
I would use the default ones, but library://video/tvshows/titles.xml are also including empty shows and it's ignoring the Kodi setting "show empty TV shows".

Comment (ronie) [reply script.embuary.helper/resources/lib/utils.py]: > Why shouldn't this be allowed?

as mentioned in the wiki https://kodi.wiki/view/Add-on_rules#Requirements_for_scripts_and_plugins

`add-ons should store all their data in their own subfolder inside the addon_data directory. Access (read/write/delete) to any other files / folders is not allowed by default.`

if these custom nodes are used to overcome certain issues in kodi, as i gather from your reply, i'd much rather see some effort the address these issues in core instead of introducing skin/script workarounds.

Comment (sualfred) [reply script.embuary.helper/resources/lib/utils.py]: @ronie 
Sorry, but that would mean that the library node editor also has to be removed. It does exactly the same and stores nothing in the addon_data folder, because library nodes only work in library:// and not outside of the path. If that would be the case skins could store them directly in the skin directory.
This function just offers the feature that skins can ship premade nodes without pointing users to https://kodi.wiki/view/Video_nodes or the https://kodi.wiki/view/Add-on:Library_Node_Editor which are too complicated for most of the common users.  It simplify things. And creates advantages to script.skinshortcuts, because these nodes can be listed there automatically.

My example only showed one scenario that is helpful as workaround but it can do a lot more. And sorry, I know nearly nothing about c++. I'm a skinner who just learned Python. Don't expect that I'm able to learn Chinese too. I don't have any single time left for it. As much as I would do it, but my pregnant wife and my daughter would kill me.

Comment (enen92) [reply script.embuary.helper/resources/lib/utils.py]: We sometimes make exceptions. For instance, add-ons cannot write to the keymap.xml file but the keymap editor add-on does it and is the only add-on allowed to do it. It is not a question of removing add-ons that do the same...but more of a compromise to reduce the error footprint while not compromising functionality. If fixing core is not an option maybe call into the node editor add-on?

Comment (sualfred) [reply script.embuary.helper/resources/lib/utils.py]: Doesn't help and won't work. 

Don't get me wrong. I agree that addons should store everything ins it's own addon_data folder when it's possible. But things like your mentioned keymap, or a profile based custom node, which can the user create on his own if he understands what he is doing(!!!!), that should be allowed to be handled, or in this case extended, by an addon. I do not harm any default library node. I also force a skin folder name, so it's clear and structured.

C'mon guys... don't let me get frustrated again because of a rule that cannot fit all scenarios.  We are talking just about node.xmls's not about a hacky core thing. It makes it more userfriendly and skinners don't have to deal with annoyning questions of unexperienced users and low WAF.

Comment (ronie) [reply script.embuary.helper/resources/lib/utils.py]: in case of the library node editor / keymap editor:
- it is manually installed by the user
- the user knowingly and willingly uses it to modify those files.
that's a use-case we've made an exception for.

as for your script, it is auto-installed and modifications are made in the background without the user realizing changes are made to his system.
in addition, those copied library nodes will be left behind when a user decides to uninstall a skin (or your helper addon).

btw, it wasn't my intention to say that _you_ should be the one to take on the job of fixing kodi core. but i'd rather have addon devs create a bug report on github than seeing them implement such workarounds.

Comment (sualfred) [reply script.embuary.helper/resources/lib/utils.py]: It's not a bug. It's a common and mostly correct behaviour. If I back out to the home screen and then return to the video DB it's useful to have the window history there. There is just no way to override this. And there are scenarios where those things just don't fit. 
I also wouldn't call it a workaround. It's a feature in my eyes that can make skinners life easier for other scenarios like shipping own playlists, what already works.

Kodi is not really self maintaining. The textures stuff get huge after months of runnings, addon_data is full of stuff from old addons or deleted addons.. I don't think 1kb big files will harm anybody next to all the other dead leftovers. If it calms you, I can put in an cleanup option to delete all script created nodes (but I bet noone will ever use it :wink: )

And with your argument that users are knowingly willing to modify files -> I don't modify anything. I copy possible new filter ones. In Embuary exactly 2 nodes.  And the script does have options to control everything: Kodi settings, addon settings, but  I let users know and decide. I show very big YES/NO dialog what I want to setup everything correctly so everything would work right out of the box (see https://github.com/sualfred/script.embuary.helper/wiki/Script:-Change-global-Kodi-settings#attention-and-warning ). 
Simplified and just-working-out-the-box > complicated and manual setup with a higher frustating level and jump-off-rate. It's all about the user experience.

Long story short:
Without this Python function my TV show hub integration will "break" and people will get empty TV shows again. 

Anyway.. you have to power to approve it. Not me. But I hope you can understand that I cannot agree and understand this rule for this case.

**Comment by sualfred** [comment]
@ronie @enen92 

I've remove the skinnodes function and replaced it with a tvshowsroot() for my own needs.

**Comment by sualfred** [comment]
Found a way to force the reset of the path history via skin level. No need for any workaround via Python or skin provided nodes.

`ActivateWindow(videos,videodb://tvshows/titles/?force_root=$INFO[System.Time(mm)]$INFO[System.Time(ss)],return)`

**Comment by enen92** [comment]
Awesome, thanks a lot for adjusting
